Cacheless Data Download

pickles-bread-and-butter opened the pull request:
## Overview:

Adds a cacheless data grabbing by passing only the locations of the file on the filesystem and not downloading it to the local cache first.

## Adds:

1. New class ColumnBytesFileReader that reads the data directly from the column bytes file location
2. Moves common logic into this class
3. Existing ColumnBytesFileCache inherits and overwrites the read to copy the file to local system first
4. Changes the Visitor to the data row to take a reading callable instead of the class, that callable now take the column byte location object and reads and returns

## Testing:

1. Done here in the CI
2. Will add more internal testing to the internal docs

convexquad commented on wicker/core/column_files.py:
```diff
@@ -182,16 +183,49 @@ def _write_column(self, column_name: str) -> None:
+        example: validation.AvroRecord,
+        schema: schema.DatasetSchema,
+    ):
+        visitor = ResolvePointersVisitor(example, schema, lambda cbf_info: self.read(cbf_info))
```
@isaak-willett I have a preference to actually pass the `ColumnBytesFileReader` (i.e. `self`) to the `ResolvePointersVisitor` as its last argument (instead of passing a `lambda`). Then the visitor can just call the `read` function on the reader.

pickles-bread-and-butter replied:
Done

convexquad commented on wicker/core/datasets.py:
```diff
@@ -68,15 +75,20 @@ def __init__(
         self._treat_objects_as_bytes = treat_objects_as_bytes
 
-        self._column_bytes_file_cache: Optional[ColumnBytesFileCache] = None
+        self._column_bytes_file_reader: ColumnBytesFileReader = ColumnBytesFileReader(
```
It would be nice to create this in the else condition of `if local_cache_path_prefix is not None:` rather than creating and then possibly re-creating `self._column_bytes_file_reader`.

pickles-bread-and-butter replied:
Done

convexquad reviewed (commented):
@isaak-willett I added one comment - please take a look. Then, do your best to test local storage both with and without the cache data download.

Once we are ready to merge, I think the best thing is that we merge this PR into your branch `feature/isaak/local_fs_dataset` from your last PR and then merge your last PR into `master`.

pickles-bread-and-butter commented:
> @isaak-willett I added one comment - please take a look. Then, do your best to test local storage both with and without the cache data download.
> 
> Once we are ready to merge, I think the best thing is that we merge this PR into your branch `feature/isaak/local_fs_dataset` from your last PR and then merge your last PR into `master`.

@convexquad I added a test that ensures that the cache isn't being written to when there is no prefix passed. Check the `test_datasets` file and look at the cached vs non-cached `FileSystemDataset` tests. There is now a check on the dir size of the COLUMN_CONCATENATED_FILE dir for both the cache and the original dir.

In the non cache the test asserts that folder is 0 bytes and the cached asserts that the col files copied there are identical to the original written dir.

pickles-bread-and-butter commented on tests/test_datasets.py:
```diff
@@ -78,23 +92,54 @@ def _setup_storage(self) -> Iterator[Tuple[FakeFileSystemDataStorage, WickerPath
+                # test that nothing got copied to the non passed cache dir
+                tmpdir_size = get_size(tmp_cache_dir)
+                # assert that nothing is copied to folder that would have been used as cache
+                assert tmpdir_size == 0
```
@convexquad here is the non cache assert

pickles-bread-and-butter commented on tests/test_datasets.py:
```diff
@@ -78,23 +92,54 @@ def _setup_storage(self) -> Iterator[Tuple[FakeFileSystemDataStorage, WickerPath
+                tmpdir_size = get_size(tmp_cache_dir)
+                expect_dir_size = get_size(os.path.join(tmpdir, "__COLUMN_CONCATENATED_FILES__"))
+                # test that all column files are copied over that are meant to be
+                assert tmpdir_size == expect_dir_size
```
@convexquad here is the cache assertion

pickles-bread-and-butter commented on CODEOWNERS:
```diff
@@ -1 +1 @@
-*       @aalavian @anantsimran @chrisochoatri @isaak-willett @marccarre
+*       @aalavian @anantsimran @chrisochoatri @convexquad @pickles-bread-and-butter @marccarre
```
Had to change my username on gh so I need to update myself

pickles-bread-and-butter commented:
Going to merge this one in to the other
